pkg/certificate/mrc_reconciler.go

```diff
@@ -0,0 +1,161 @@
+
+	// UseCaseBootstrap maps to the certificate use case for the  Bootstrap secrets
+	UseCaseBootstrap UseCase = "bootstrap"
+
+	// UseCaseGateway maps to the certificate use case for the  Gateway ingress cert.
+	UseCaseGateway UseCase = "gateway"
+)
+
+func (uc UseCase) String() string {
+	return string(uc)
+}
+func (m *Manager) handleMRCEvent(event MRCEvent) error {
+	mrc := event.MRC
+	// TODO(5046): remove this first call to setIssuers.
+	if err := m.setIssuers(event.MRC); err != nil {
+		return err
+	}
+	// TODO(5046): improve logging in this method.
+	if shouldUpdateMRCComponentStatus(mrc) {
+		if err := m.updateMRCComponentStatus(mrc); err != nil {
+			return err
+		}
+	} else if !isTerminal(mrc) {
+		if err := m.updateMRCState(mrc); err != nil {
+			return err
+		}
+	} else if m.shouldSetIssuers(mrc) {
+		log.Info().Msgf("setting new certificate issuers on the Certificate Manager for MRC %s in stage %s", mrc.GetName(), mrc.Status.State)
+		return m.setIssuers(mrc)
+	}
+	return nil
+}
+
+// Component statuses determine if we're ready to transition the mrc to the next state.
+// TODO(5046): add unit tests for this, and other methods in here.
+func shouldUpdateMRCComponentStatus(mrc *v1alpha2.MeshRootCertificate) bool {
+	if isTerminal(mrc) {
+		return false
+	}
+
+	if len(mrc.Status.Conditions) == 0 {
+		return false
+	}
+
+	transitionAfter := mrc.Status.TransitionAfter
+	if transitionAfter == nil {
+		return false
+	}
+
+	//The truncate is needed so we truncate the monotonic clock.
+	return transitionAfter.Truncate(0).After(time.Now())
+}
+
+// update the MRCComponentStatus
+func (m *Manager) updateMRCComponentStatus(mrc *v1alpha2.MeshRootCertificate) error {
+	// TODO(5046): determine the status to set the owned components to.
+	var status v1alpha2.MeshRootCertificateComponentStatus
+
+	for _, useCase := range m.ownedUseCases {
+		switch useCase {
+		case UseCaseValidatingWebhook:
+			// validating, issuing unknown.
+			mrc.Status.ComponentStatuses.ValidatingWebhook = status
+		case UseCaseMutatingWebhook:
+			mrc.Status.ComponentStatuses.MutatingWebhook = status
+		case UseCaseXDSControlPlane:
+			mrc.Status.ComponentStatuses.XDSControlPlane = status
+		case UseCaseSidecar:
+			mrc.Status.ComponentStatuses.Sidecar = status
+		case UseCaseBootstrap:
+			mrc.Status.ComponentStatuses.Bootstrap = status
+		case UseCaseGateway:
+			mrc.Status.ComponentStatuses.Gateway = status
+		}
+	}
+	return m.mrcClient.UpdateMeshRootCertificate(mrc)
+}
+
+// isTerminal checks if the MRC is in a terminal state.
+func isTerminal(mrc *v1alpha2.MeshRootCertificate) bool {
+	// TODO(5046): check if the mrc is in a terminal state.
+	return true
+}
+
+func (m *Manager) updateMRCState(mrc *v1alpha2.MeshRootCertificate) error {
+	// TODO(5046): update the MRC state & conditions, and issue the update via the MRCClient.
+	// add the retry loop as well.
+
+	// If it's not in a terminal state, set the next transition time.
+	if isTerminal(mrc) {
```

Comment:
I don't know that `transitionAfter` is tied to terminal states like this. Please refer to the [gist](https://gist.github.com/keithmattix/9ac73abbcb5721b0ce58102ac3ebff29)

Reply:
there would be no next transition time if it's in a terminal state though, no? since there's nothing to transition to

Reply:
ut this blindly adds a transition time even if we haven't determined that componentStatus needs to update based on intent and/or conditions. One of updateMRCState or shouldUpdateMRCComponentStatus needs a TODO to add conditional logic based on the gist (probably the latter function)

Reply:
why does it do it blindly? that's not called out anywhere so I don't see the need to make that decision. Not setting a transitionTime if we know there's no future transition makes more sense to me

Reply:
I'm saying the code here blindly sets the transition. As it stands, there's no place in the PR for osm-controller to check that, e.g. `spec.intent == Passive && status.Conditions[type=Accepted].Status == True && status.Conditions[type=ValidatingRollout] => { x.Status == False && x.Reason == Pending }` on the MRC in order to set the transition time